File: FestivalJava/Networking/src/main/java/protobuf/ProtoFestivalProxy.java
package protobuf;

import AppServices.IFestivalClient;
import AppServices.IFestivalServer;
import Domain.Artist;
import Domain.Show;
import Domain.User;
import Validation.Exceptions.ServiceException;
import Validation.Exceptions.ValidatorException;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * Created by Sebi on 25-Apr-17.
 */
public class ProtoFestivalProxy implements IFestivalServer {
    private String host;
    private int port;

    private IFestivalClient client;

    private InputStream inputStream;
    private OutputStream outputStream;
    private Socket connection;

    private BlockingQueue<FestivalProtobufs.FestivalResponse> qresponses;
    private volatile boolean finished;

    public ProtoFestivalProxy(String host, int port){
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingQueue<>();
    }

    private void initializeConnection(){
        try{
            connection = new Socket(host, port);
            outputStream = connection.getOutputStream();
            inputStream = connection.getInputStream();
            finished = false;
            startReader();
        } catch (IOException e){
            e.printStackTrace();
        }
    }

    private void closeConnection(){
        finished = true;
        try {
            inputStream.close();
            outputStream.close();
            connection.close();
            client = null;
        } catch (IOException e){
            e.printStackTrace();
        }
    }

    private void startReader(){
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }

    private class ReaderThread implements Runnable {
        public void run(){
            while (!finished){
                try{
                    FestivalProtobufs.FestivalResponse response =
                            FestivalProtobufs.FestivalResponse.parseDelimitedFrom(inputStream);
                    if (isUpdate(response.getType())){

                    } else {
                        try {
                            qresponses.put(response);
                        } catch (InterruptedException e){
                            e.printStackTrace();
                        }
                    }
                } catch (IOException e){
                    System.out.println("Client reader stopped");
                    finished = true;
                }
            }
        }
    }

    private boolean isUpdate(FestivalProtobufs.FestivalResponse.Type type){
        if (type == FestivalProtobufs.FestivalResponse.Type.ShowUpdated){
            return true;
        } else {
            return false;
        }
    }

    private void sendRequest(FestivalProtobufs.FestivalRequest request){
        try{
            System.out.println("Sending request" + request);
            request.writeDelimitedTo(outputStream);
            outputStream.flush();
            System.out.println("Request sent");
        } catch (IOException e){
            e.printStackTrace();
        }
    }

    private FestivalProtobufs.FestivalResponse readResponse(){
        FestivalProtobufs.FestivalResponse response = null;

        try{
            response = qresponses.take();
        } catch (InterruptedException e){
            e.printStackTrace();
        }

        return response;
    }

    @Override
    public List<Artist> getArtists() throws ServiceException {
        sendRequest(ProtoUtils.createGetArtistsRequest());
        FestivalProtobufs.FestivalResponse response = readResponse();
        if (response.getType() == FestivalProtobufs.FestivalResponse.Type.Error){
            throw new ServiceException(response.getError());
        }
        return ProtoUtils.getArtists(response);
    }

    @Override
    public List<Show> getShowsForArtist(Integer idArtist) throws ServiceException {
        sendRequest(ProtoUtils.createGetShowsForArtistRequest(idArtist));
        FestivalProtobufs.FestivalResponse response = readResponse();
        if (response.getType() == FestivalProtobufs.FestivalResponse.Type.Error){
            throw new ServiceException(response.getError());
        }
        return ProtoUtils.getShows(response);
    }

    @Override
    public List<Show> getShowsForDate(String date) throws ServiceException {
        sendRequest(ProtoUtils.createGetShowsForDateRequest(date));
        FestivalProtobufs.FestivalResponse response = readResponse();
        if (response.getType() == FestivalProtobufs.FestivalResponse.Type.Error){
            throw new ServiceException(response.getError());
        }
        return ProtoUtils.getShows(response);
    }

    @Override
    public void buyTicketsForShow(Integer idShow, String clientName, Integer numberOfTickets, String username) throws ServiceException, ValidatorException {

    }

    @Override
    public void login(User user, IFestivalClient client) throws ServiceException {
        initializeConnection();
        sendRequest(ProtoUtils.createLoginRequest(user));
        FestivalProtobufs.FestivalResponse response = readResponse();
        if (response.getType() == FestivalProtobufs.FestivalResponse.Type.Ok){
            this.client = client;
            return;
        }
        if (response.getType() == FestivalProtobufs.FestivalResponse.Type.Error){
            throw new ServiceException(response.getError());
        }
    }

    @Override
    public void logout(User user, IFestivalClient client) throws ServiceException {
        sendRequest(ProtoUtils.createLogoutRequest(user));
        FestivalProtobufs.FestivalResponse response = readResponse();
        closeConnection();
        if (response.getType() == FestivalProtobufs.FestivalResponse.Type.Error){
            throw new ServiceException(response.getError());
        }
    }
}

File: FestivalC#/Festival/Networking/protobuff/ProtoFestivalProxy.cs
using Festival.Service;
using Festival.Service.AppServices;
using Google.Protobuf;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Festival.Model;
using Services.Validation.Exceptions;

namespace Networking.protobuff
{
    public class ProtoFestivalProxy : IFestivalServer
    {
        private string host;
        private int port;
        private NetworkStream stream;
        private IFormatter formatter;
        private TcpClient connection;

        private Queue<Protobuf.FestivalResponse> responses;
        private volatile bool finished;
        private EventWaitHandle waitHandle;

        private IFestivalClient client;

        public ProtoFestivalProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new Queue<Protobuf.FestivalResponse>();
        }

        private void sendRequest(Protobuf.FestivalRequest request)
        {
            try
            {
                request.WriteDelimitedTo(stream);
                stream.Flush();
            } catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        private Protobuf.FestivalResponse readResponse()
        {
            Protobuf.FestivalResponse response = null;
            try
            {
                waitHandle.WaitOne();
                lock (responses)
                {
                    response = responses.Dequeue();
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            return response;

        }

        private void initializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished = false;
                waitHandle = new AutoResetEvent(false);
                startReader();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        private void closeConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                waitHandle.Close();
                client = null;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        private void startReader()
        {
            Thread tw = new Thread(run);
            tw.Start();
        }

        private bool isUpdate(Protobuf.FestivalResponse response)
        {
            if (response.Type == Protobuf.FestivalResponse.Types.Type.ShowUpdated)
            {
                return true;
            }
            return false;
        }

        //background reader thread run method
        public virtual void run()
        {
            while (!finished)
            {
                try
                {
                    Protobuf.FestivalResponse response = Protobuf.FestivalResponse.Parser.ParseDelimitedFrom(stream);

                    Console.WriteLine("Response received " + response);
                    if (isUpdate(response))
                    {
                        handleUpdate(response);
                    }
                    else
                    {
                        lock (responses)
                        {
                            responses.Enqueue(response);
                        }
                        waitHandle.Set();
                    }
                }
                catch (Exception ex)
                {
                    
                }
            }
        }

        private void handleUpdate(Protobuf.FestivalResponse response)
        {
           
        }

        public List<Artist> getArtists()
        {
            sendRequest(ProtoUtils.createGetArtistsRequest());
            Protobuf.FestivalResponse response = readResponse();
            if (response.Type == Protobuf.FestivalResponse.Types.Type.Error)
            {
                throw new ServiceException(response.Error);
            }
            return ProtoUtils.getArtists(response);
        }

        public List<Show> getShowsForArtist(int idArtist)
        {
            sendRequest(ProtoUtils.createGetShowsForArtistRequest(idArtist));
            Protobuf.FestivalResponse response = readResponse();
            if (response.Type == Protobuf.FestivalResponse.Types.Type.Error)
            {
                throw new ServiceException(response.Error);
            }
            return ProtoUtils.getShows(response);
        }

        public List<Show> getShowsForDate(string date)
        {
            sendRequest(ProtoUtils.createGetShowsForDateRequest(date));
            Protobuf.FestivalResponse response = readResponse();
            if (response.Type == Protobuf.FestivalResponse.Types.Type.Error)
            {
                throw new ServiceException(response.Error);
            }
            return ProtoUtils.getShows(response);
        }

        public void buyTicketsForShow(int idShow, string clientName, int numberOfTickets, string username)
        {
            throw new NotImplementedException();
        }

        public void login(User user, IFestivalClient client)
        {
            initializeConnection();
            sendRequest(ProtoUtils.createLoginRequest(user));
            Protobuf.FestivalResponse response = readResponse();
            if (response.Type == Protobuf.FestivalResponse.Types.Type.Ok)
            {
                this.client = client;
                return;
            }
            if (response.Type == Protobuf.FestivalResponse.Types.Type.Error)
            {
                throw new ServiceException(response.Error);
            }
        }

        public void logout(User user, IFestivalClient client)
        {
            sendRequest(ProtoUtils.createLogoutRequest(user));
            Protobuf.FestivalResponse response = readResponse();
            closeConnection();
            if (response.Type == Protobuf.FestivalResponse.Types.Type.Error)
            {
                throw new ServiceException(response.Error);
            }
        }
    }
}
